src/collectd.conf.5
.\" Automatically generated by Pod::Man 2.22 (Pod::Simple 3.07)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "COLLECTD.CONF 5"
.TH COLLECTD.CONF 5 "2013-04-10" "5.3.0" "collectd"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
collectd.conf \- Configuration for the system statistics collection daemon \fBcollectd\fR
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 3
\&  BaseDir "/path/to/data/"
\&  PIDFile "/path/to/pidfile/collectd.pid"
\&  Server  "123.123.123.123" 12345
\&  
\&  LoadPlugin cpu
\&  LoadPlugin load
\&  
\&  <LoadPlugin df>
\&    Interval 3600
\&  </LoadPlugin>
\&  
\&  LoadPlugin ping
\&  <Plugin ping>
\&    Host "example.org"
\&    Host "provider.net"
\&  </Plugin>
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
This config file controls how the system statistics collection daemon
\&\fBcollectd\fR behaves. The most significant option is \fBLoadPlugin\fR, which
controls which plugins to load. These plugins ultimately define collectd's
behavior.
.PP
The syntax of this config file is similar to the config file of the famous
\&\fIApache\fR webserver. Each line contains either an option (a key and a list of
one or more values) or a section-start or \-end. Empty lines and everything
after a non-quoted hash-symbol (\f(CW\*(C`#\*(C'\fR) is ignored. \fIKeys\fR are unquoted
strings, consisting only of alphanumeric characters and the underscore (\f(CW\*(C`_\*(C'\fR)
character. Keys are handled case insensitive by \fIcollectd\fR itself and all
plugins included with it. \fIValues\fR can either be an \fIunquoted string\fR, a
\&\fIquoted string\fR (enclosed in double-quotes) a \fInumber\fR or a \fIboolean\fR
expression. \fIUnquoted strings\fR consist of only alphanumeric characters and
underscores (\f(CW\*(C`_\*(C'\fR) and do not need to be quoted. \fIQuoted strings\fR are
enclosed in double quotes (\f(CW\*(C`"\*(C'\fR). You can use the backslash character (\f(CW\*(C`\e\*(C'\fR)
to include double quotes as part of the string. \fINumbers\fR can be specified in
decimal and floating point format (using a dot \f(CW\*(C`.\*(C'\fR as decimal separator),
hexadecimal when using the \f(CW\*(C`0x\*(C'\fR prefix and octal with a leading zero (\f(CW0\fR).
\&\fIBoolean\fR values are either \fBtrue\fR or \fBfalse\fR.
.PP
Lines may be wrapped by using \f(CW\*(C`\e\*(C'\fR as the last character before the newline.
This allows long lines to be split into multiple lines. Quoted strings may be
wrapped as well. However, those are treated special in that whitespace at the
beginning of the following lines will be ignored, which allows for nicely
indenting the wrapped lines.
.PP
The configuration is read and processed in order, i.e. from top to bottom. So
the plugins are loaded in the order listed in this config file. It is a good
idea to load any logging plugins first in order to catch messages from plugins
during configuration. Also, the \f(CW\*(C`LoadPlugin\*(C'\fR option \fBmust\fR occur \fBbefore\fR
the appropriate \f(CW\*(C`<Plugin ...>\*(C'\fR block.
.SH "GLOBAL OPTIONS"
.IX Header "GLOBAL OPTIONS"
.IP "\fBBaseDir\fR \fIDirectory\fR" 4
.IX Item "BaseDir Directory"
Sets the base directory. This is the directory beneath all RRD-files are
created. Possibly more subdirectories are created. This is also the working
directory for the daemon.
.IP "\fBLoadPlugin\fR \fIPlugin\fR" 4
.IX Item "LoadPlugin Plugin"
Loads the plugin \fIPlugin\fR. There must be at least one such line or \fBcollectd\fR
will be mostly useless.
.Sp
Starting with collectd 4.9, this may also be a block in which further options
affecting the behavior of \fBLoadPlugin\fR may be specified. The following
options are allowed inside a \fBLoadPlugin\fR block:
.Sp
.Vb 4
\&  <LoadPlugin perl>
\&    Globals true
\&    Interval 10
\&  </LoadPlugin>
.Ve
.RS 4
.IP "\fBGlobals\fR \fBtrue|false\fR" 4
.IX Item "Globals true|false"
If enabled, collectd will export all global symbols of the plugin (and of all
libraries loaded as dependencies of the plugin) and, thus, makes those symbols
available for resolving unresolved symbols in subsequently loaded plugins if
that is supported by your system.
.Sp
This is useful (or possibly even required), e.g., when loading a plugin that
embeds some scripting language into the daemon (e.g. the \fIPerl\fR and
\&\fIPython plugins\fR). Scripting languages usually provide means to load
extensions written in C. Those extensions require symbols provided by the
interpreter, which is loaded as a dependency of the respective collectd plugin.
See the documentation of those plugins (e.g., \fIcollectd\-perl\fR\|(5) or
\&\fIcollectd\-python\fR\|(5)) for details.
.Sp
By default, this is disabled. As a special exception, if the plugin name is
either \f(CW\*(C`perl\*(C'\fR or \f(CW\*(C`python\*(C'\fR, the default is changed to enabled in order to keep
the average user from ever having to deal with this low level linking stuff.
.IP "\fBInterval\fR \fISeconds\fR" 4
.IX Item "Interval Seconds"
Sets a plugin-specific interval for collecting metrics. This overrides the
global \fBInterval\fR setting. If a plugin provides own support for specifying an
interval, that setting will take precedence.
.RE
.RS 4
.RE
.IP "\fBInclude\fR \fIPath\fR [\fIpattern\fR]" 4
.IX Item "Include Path [pattern]"
If \fIPath\fR points to a file, includes that file. If \fIPath\fR points to a
directory, recursively includes all files within that directory and its
subdirectories. If the \f(CW\*(C`wordexp\*(C'\fR function is available on your system,
shell-like wildcards are expanded before files are included. This means you can
use statements like the following:
.Sp
.Vb 1
\&  Include "/etc/collectd.d/*.conf"
.Ve
.Sp
If the \f(CW\*(C`fnmatch\*(C'\fR function is available on your system, a shell-like wildcard
\&\fIpattern\fR may be specified to filter which files to include. This may be used
in combination with recursively including a directory to easily be able to
arbitrarily mix configuration files and other documents (e.g. \s-1README\s0 files).
The following statement is similar to the example above but includes all files
matching \f(CW\*(C`*.conf\*(C'\fR in any subdirectory of \f(CW\*(C`/etc/collectd.d\*(C'\fR:
.Sp
.Vb 1
\&  Include "/etc/collectd.d" "*.conf"
.Ve
.Sp
If more than one files are included by a single \fBInclude\fR option, the files
will be included in lexicographical order (as defined by the \f(CW\*(C`strcmp\*(C'\fR
function). Thus, you can e.\ g. use numbered prefixes to specify the
order in which the files are loaded.
.Sp
To prevent loops and shooting yourself in the foot in interesting ways the
nesting is limited to a depth of 8\ levels, which should be sufficient for
most uses. Since symlinks are followed it is still possible to crash the daemon
by looping symlinks. In our opinion significant stupidity should result in an
appropriate amount of pain.
.Sp
It is no problem to have a block like \f(CW\*(C`<Plugin foo>\*(C'\fR in more than one
file, but you cannot include files from within blocks.
.IP "\fBPIDFile\fR \fIFile\fR" 4
.IX Item "PIDFile File"
Sets where to write the \s-1PID\s0 file to. This file is overwritten when it exists
and deleted when the program is stopped. Some init-scripts might override this
setting using the \fB\-P\fR command-line option.
.IP "\fBPluginDir\fR \fIDirectory\fR" 4
.IX Item "PluginDir Directory"
Path to the plugins (shared objects) of collectd.
.IP "\fBTypesDB\fR \fIFile\fR [\fIFile\fR ...]" 4
.IX Item "TypesDB File [File ...]"
Set one or more files that contain the data-set descriptions. See
\&\fItypes.db\fR\|(5) for a description of the format of this file.
.IP "\fBInterval\fR \fISeconds\fR" 4
.IX Item "Interval Seconds"
Configures the interval in which to query the read plugins. Obviously smaller
values lead to a higher system load produced by collectd, while higher values
lead to more coarse statistics.
.Sp
\&\fBWarning:\fR You should set this once and then never touch it again. If you do,
\&\fIyou will have to delete all your \s-1RRD\s0 files\fR or know some serious RRDtool
magic! (Assuming you're using the \fIRRDtool\fR or \fIRRDCacheD\fR plugin.)
.IP "\fBTimeout\fR \fIIterations\fR" 4
.IX Item "Timeout Iterations"
Consider a value list \*(L"missing\*(R" when no update has been read or received for
\&\fIIterations\fR iterations. By default, \fIcollectd\fR considers a value list
missing when no update has been received for twice the update interval. Since
this setting uses iterations, the maximum allowed time without update depends
on the \fIInterval\fR information contained in each value list. This is used in
the \fIThreshold\fR configuration to dispatch notifications about missing values,
see \fIcollectd\-threshold\fR\|(5) for details.
.IP "\fBReadThreads\fR \fINum\fR" 4
.IX Item "ReadThreads Num"
Number of threads to start for reading plugins. The default value is \fB5\fR, but
you may want to increase this if you have more than five plugins that take a
long time to read. Mostly those are plugins that do network-IO. Setting this to
a value higher than the number of registered read callbacks is not recommended.
.IP "\fBWriteThreads\fR \fINum\fR" 4
.IX Item "WriteThreads Num"
Number of threads to start for dispatching value lists to write plugins. The
default value is \fB5\fR, but you may want to increase this if you have more than
five plugins that may take relatively long to write to.
.IP "\fBHostname\fR \fIName\fR" 4
.IX Item "Hostname Name"
Sets the hostname that identifies a host. If you omit this setting, the
hostname will be determined using the \fIgethostname\fR\|(2) system call.
.IP "\fBFQDNLookup\fR \fBtrue|false\fR" 4
.IX Item "FQDNLookup true|false"
If \fBHostname\fR is determined automatically this setting controls whether or not
the daemon should try to figure out the \*(L"fully qualified domain name\*(R", \s-1FQDN\s0.
This is done using a lookup of the name returned by \f(CW\*(C`gethostname\*(C'\fR. This option
is enabled by default.
.IP "\fBPreCacheChain\fR \fIChainName\fR" 4
.IX Item "PreCacheChain ChainName"
.PD 0
.IP "\fBPostCacheChain\fR \fIChainName\fR" 4
.IX Item "PostCacheChain ChainName"
.PD
Configure the name of the \*(L"pre-cache chain\*(R" and the \*(L"post-cache chain\*(R". Please
see \*(L"\s-1FILTER\s0 \s-1CONFIGURATION\s0\*(R" below on information on chains and how these
setting change the daemon's behavior.
.SH "PLUGIN OPTIONS"
.IX Header "PLUGIN OPTIONS"
Some plugins may register own options. These options must be enclosed in a
\&\f(CW\*(C`Plugin\*(C'\fR\-Section. Which options exist depends on the plugin used. Some plugins
require external configuration, too. The \f(CW\*(C`apache plugin\*(C'\fR, for example,
required \f(CW\*(C`mod_status\*(C'\fR to be configured in the webserver you're going to
collect data from. These plugins are listed below as well, even if they don't
require any configuration within collectd's configuration file.
.PP
A list of all plugins and a short summary for each plugin can be found in the
\&\fI\s-1README\s0\fR file shipped with the sourcecode and hopefully binary packets as
well.
.ie n .SS "Plugin ""aggregation"""
.el .SS "Plugin \f(CWaggregation\fP"
.IX Subsection "Plugin aggregation"
The \fIAggregation plugin\fR makes it possible to aggregate several values into
one using aggregation functions such as \fIsum\fR, \fIaverage\fR, \fImin\fR and \fImax\fR.
This can be put to a wide variety of uses, e.g. average and total \s-1CPU\s0
statistics for your entire fleet.
.PP
The grouping is powerful but, as with many powerful tools, may be a bit
difficult to wrap your head around. The grouping will therefore be
demonstrated using an example: The average and sum of the \s-1CPU\s0 usage across
all CPUs of each host is to be calculated.
.PP
To select all the affected values for our example, set \f(CW\*(C`Plugin cpu\*(C'\fR and
\&\f(CW\*(C`Type cpu\*(C'\fR. The other values are left unspecified, meaning \*(L"all values\*(R". The
\&\fIHost\fR, \fIPlugin\fR, \fIPluginInstance\fR, \fIType\fR and \fITypeInstance\fR options
work as if they were specified in the \f(CW\*(C`WHERE\*(C'\fR clause of an \f(CW\*(C`SELECT\*(C'\fR \s-1SQL\s0
statement.
.PP
.Vb 2
\&  Plugin "cpu"
\&  Type "cpu"
.Ve
.PP
Although the \fIHost\fR, \fIPluginInstance\fR (\s-1CPU\s0 number, i.e. 0, 1, 2, ...)  and
\&\fITypeInstance\fR (idle, user, system, ...) fields are left unspecified in the
example, the intention is to have a new value for each host / type instance

NEWS
Test mit Login `niki'


src/collectd.c
/**
 * collectd - src/collectd.c
 * Copyright (C) 2005-2007  Florian octo Forster
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; only version 2 of the License is applicable.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
 *
 * Authors:
 *   Florian octo Forster <octo at verplant.org>
 *   Alvaro Barcellos <alvaro.barcellos at gmail.com>
 **/

#include "collectd.h"
#include "common.h"

#include <sys/types.h>
#include <sys/socket.h>
#include <netdb.h>

#include <pthread.h>

#include "plugin.h"
#include "configfile.h"

#if HAVE_STATGRAB_H
# include <statgrab.h>
#endif

/*
 * Global variables
 */
char hostname_g[DATA_MAX_NAME_LEN];
cdtime_t interval_g;
int  timeout_g;
#if HAVE_LIBKSTAT
kstat_ctl_t *kc;
#endif /* HAVE_LIBKSTAT */

static int loop = 0;

static void *do_flush (void __attribute__((unused)) *arg)
{
	INFO ("Flushing all data.");
	plugin_flush (/* plugin = */ NULL,
			/* timeout = */ 0,
			/* ident = */ NULL);
	INFO ("Finished flushing all data.");
	pthread_exit (NULL);
	return NULL;
}

static void sig_int_handler (int __attribute__((unused)) signal)
{
	loop++;
}

static void sig_term_handler (int __attribute__((unused)) signal)
{
	loop++;
}

static void sig_usr1_handler (int __attribute__((unused)) signal)
{
	pthread_t      thread;
	pthread_attr_t attr;

	/* flushing the data might take a while,
	 * so it should be done asynchronously */
	pthread_attr_init (&attr);
	pthread_attr_setdetachstate (&attr, PTHREAD_CREATE_DETACHED);
	pthread_create (&thread, &attr, do_flush, NULL);
	pthread_attr_destroy (&attr);
}

static int init_hostname (void)
{
	const char *str;

	struct addrinfo  ai_hints;
	struct addrinfo *ai_list;
	struct addrinfo *ai_ptr;
	int status;

	str = global_option_get ("Hostname");
	if (str != NULL)
	{
		sstrncpy (hostname_g, str, sizeof (hostname_g));
		return (0);
	}

	if (gethostname (hostname_g, sizeof (hostname_g)) != 0)
	{
		fprintf (stderr, "`gethostname' failed and no "
				"hostname was configured.\n");
		return (-1);
	}

	str = global_option_get ("FQDNLookup");
	if (IS_FALSE (str))
		return (0);

	memset (&ai_hints, '\0', sizeof (ai_hints));
	ai_hints.ai_flags = AI_CANONNAME;

	status = getaddrinfo (hostname_g, NULL, &ai_hints, &ai_list);
	if (status != 0)
	{
		ERROR ("Looking up \"%s\" failed. You have set the "
				"\"FQDNLookup\" option, but I cannot resolve "
				"my hostname to a fully qualified domain "
				"name. Please fix you network "
				"configuration.", hostname_g);
		return (-1);
	}

	for (ai_ptr = ai_list; ai_ptr != NULL; ai_ptr = ai_ptr->ai_next)
	{
		if (ai_ptr->ai_canonname == NULL)
			continue;

		sstrncpy (hostname_g, ai_ptr->ai_canonname, sizeof (hostname_g));
		break;
	}

	freeaddrinfo (ai_list);
	return (0);
} /* int init_hostname */

static int init_global_variables (void)
{
	char const *str;

	interval_g = cf_get_default_interval ();
	assert (interval_g > 0);
	DEBUG ("interval_g = %.3f;", CDTIME_T_TO_DOUBLE (interval_g));

	str = global_option_get ("Timeout");
	if (str == NULL)
		str = "2";
	timeout_g = atoi (str);
	if (timeout_g <= 1)
	{
		fprintf (stderr, "Cannot set the timeout to a correct value.\n"
				"Please check your settings.\n");
		return (-1);
	}
	DEBUG ("timeout_g = %i;", timeout_g);

	if (init_hostname () != 0)
		return (-1);
	DEBUG ("hostname_g = %s;", hostname_g);

	return (0);
} /* int init_global_variables */

static int change_basedir (const char *orig_dir)
{
	char *dir;
	size_t dirlen;
	int status;

	dir = strdup (orig_dir);
	if (dir == NULL)
	{
		char errbuf[1024];
		ERROR ("strdup failed: %s",
				sstrerror (errno, errbuf, sizeof (errbuf)));
		return (-1);
	}
	
	dirlen = strlen (dir);
	while ((dirlen > 0) && (dir[dirlen - 1] == '/'))
		dir[--dirlen] = '\0';

	if (dirlen <= 0)
		return (-1);

	status = chdir (dir);
	if (status == 0)
	{
		free (dir);
		return (0);
	}
	else if (errno != ENOENT)
	{
		char errbuf[1024];
		ERROR ("change_basedir: chdir (%s): %s", dir,
				sstrerror (errno, errbuf, sizeof (errbuf)));
		free (dir);
		return (-1);
	}

	status = mkdir (dir, S_IRWXU | S_IRWXG | S_IRWXO);
	if (status != 0)
	{
		char errbuf[1024];
		ERROR ("change_basedir: mkdir (%s): %s", dir,
				sstrerror (errno, errbuf, sizeof (errbuf)));
		free (dir);
		return (-1);
	}

	status = chdir (dir);
	if (status != 0)
	{
		char errbuf[1024];
		ERROR ("change_basedir: chdir (%s): %s", dir,
				sstrerror (errno, errbuf, sizeof (errbuf)));
		free (dir);
		return (-1);
	}

	free (dir);
	return (0);
} /* static int change_basedir (char *dir) */

#if HAVE_LIBKSTAT
static void update_kstat (void)
{
	if (kc == NULL)
	{
		if ((kc = kstat_open ()) == NULL)
			ERROR ("Unable to open kstat control structure");
	}
	else
	{
		kid_t kid;
		kid = kstat_chain_update (kc);
		if (kid > 0)
		{
			INFO ("kstat chain has been updated");
			plugin_init_all ();
		}
		else if (kid < 0)
			ERROR ("kstat chain update failed");
		/* else: everything works as expected */
	}

	return;
} /* static void update_kstat (void) */
#endif /* HAVE_LIBKSTAT */

/* TODO
 * Remove all settings but `-f' and `-C'
 */
static void exit_usage (int status)
{
	printf ("Usage: "PACKAGE" [OPTIONS]\n\n"
			
			"Available options:\n"
			"  General:\n"
			"    -C <file>       Configuration file.\n"
			"                    Default: "CONFIGFILE"\n"
			"    -t              Test config and exit.\n"
			"    -T              Test plugin read and exit.\n"
			"    -P <file>       PID-file.\n"
			"                    Default: "PIDFILE"\n"
#if COLLECT_DAEMON
			"    -f              Don't fork to the background.\n"
#endif
			"    -h              Display help (this message)\n"
			"\nBuiltin defaults:\n"
			"  Config file       "CONFIGFILE"\n"
			"  PID file          "PIDFILE"\n"
			"  Plugin directory  "PLUGINDIR"\n"
			"  Data directory    "PKGLOCALSTATEDIR"\n"
			"\n"PACKAGE" "VERSION", http://collectd.org/\n"
			"by Florian octo Forster <octo@verplant.org>\n"
			"for contributions see `AUTHORS'\n");
	exit (status);
} /* static void exit_usage (int status) */

static int do_init (void)
{
#if HAVE_LIBKSTAT
	kc = NULL;
	update_kstat ();
#endif

#if HAVE_LIBSTATGRAB
	if (sg_init ())
	{
		ERROR ("sg_init: %s", sg_str_error (sg_get_error ()));
		return (-1);
	}

	if (sg_drop_privileges ())
	{
		ERROR ("sg_drop_privileges: %s", sg_str_error (sg_get_error ()));
		return (-1);
	}
#endif

	plugin_init_all ();

	return (0);
} /* int do_init () */


static int do_loop (void)
{
	cdtime_t interval = cf_get_default_interval ();
	cdtime_t wait_until;

	wait_until = cdtime () + interval;

	while (loop == 0)
	{
		struct timespec ts_wait = { 0, 0 };
		cdtime_t now;

#if HAVE_LIBKSTAT
		update_kstat ();
#endif

		/* Issue all plugins */
		plugin_read_all ();

		now = cdtime ();
		if (now >= wait_until)
		{
			WARNING ("Not sleeping because the next interval is "
					"%.3f seconds in the past!",
					CDTIME_T_TO_DOUBLE (now - wait_until));
			wait_until = now + interval;
			continue;
		}

		CDTIME_T_TO_TIMESPEC (wait_until - now, &ts_wait);
		wait_until = wait_until + interval;

		while ((loop == 0) && (nanosleep (&ts_wait, &ts_wait) != 0))
		{
			if (errno != EINTR)
			{
				char errbuf[1024];
				ERROR ("nanosleep failed: %s",
						sstrerror (errno, errbuf,
							sizeof (errbuf)));
				return (-1);
			}
		}
	} /* while (loop == 0) */

	return (0);
} /* int do_loop */

static int do_shutdown (void)
{
	plugin_shutdown_all ();
	return (0);
} /* int do_shutdown */

#if COLLECT_DAEMON
static int pidfile_create (void)
{
	FILE *fh;
	const char *file = global_option_get ("PIDFile");

	if ((fh = fopen (file, "w")) == NULL)
	{
		char errbuf[1024];
		ERROR ("fopen (%s): %s", file,
				sstrerror (errno, errbuf, sizeof (errbuf)));
		return (1);
	}

	fprintf (fh, "%i\n", (int) getpid ());
	fclose(fh);

	return (0);
} /* static int pidfile_create (const char *file) */

static int pidfile_remove (void)
{
	const char *file = global_option_get ("PIDFile");

	DEBUG ("unlink (%s)", (file != NULL) ? file : "<null>");
	return (unlink (file));
} /* static int pidfile_remove (const char *file) */
#endif /* COLLECT_DAEMON */

int main (int argc, char **argv)
{
	struct sigaction sig_int_action;
	struct sigaction sig_term_action;
	struct sigaction sig_usr1_action;
	struct sigaction sig_pipe_action;
	char *configfile = CONFIGFILE;
	int test_config  = 0;
	int test_readall = 0;
	const char *basedir;
#if COLLECT_DAEMON
	struct sigaction sig_chld_action;
	pid_t pid;
	int daemonize    = 1;
#endif
	int exit_status = 0;

	/* read options */
	while (1)
	{
		int c;

		c = getopt (argc, argv, "htTC:"
#if COLLECT_DAEMON
				"fP:"
#endif
		);

		if (c == -1)
			break;

		switch (c)
		{
			case 'C':
				configfile = optarg;
				break;
			case 't':
				test_config = 1;
				break;
			case 'T':
				test_readall = 1;
				global_option_set ("ReadThreads", "-1");
#if COLLECT_DAEMON
				daemonize = 0;
#endif /* COLLECT_DAEMON */
				break;
#if COLLECT_DAEMON
			case 'P':
				global_option_set ("PIDFile", optarg);
				break;
			case 'f':
				daemonize = 0;
				break;
#endif /* COLLECT_DAEMON */
			case 'h':
				exit_usage (0);
				break;
			default:
				exit_usage (1);
		} /* switch (c) */
	} /* while (1) */

	if (optind < argc)
		exit_usage (1);

	plugin_init_ctx ();

	/*
	 * Read options from the config file, the environment and the command
	 * line (in that order, with later options overwriting previous ones in
	 * general).
	 * Also, this will automatically load modules.
	 */
	if (cf_read (configfile))
	{
		fprintf (stderr, "Error: Reading the config file failed!\n"
				"Read the syslog for details.\n");
		return (1);
	}

	/*
	 * Change directory. We do this _after_ reading the config and loading
	 * modules to relative paths work as expected.
	 */
	if ((basedir = global_option_get ("BaseDir")) == NULL)
	{
		fprintf (stderr, "Don't have a basedir to use. This should not happen. Ever.");
		return (1);
	}
	else if (change_basedir (basedir))
	{
		fprintf (stderr, "Error: Unable to change to directory `%s'.\n", basedir);
		return (1);
	}

	/*
	 * Set global variables or, if that failes, exit. We cannot run with
	 * them being uninitialized. If nothing is configured, then defaults
	 * are being used. So this means that the user has actually done
	 * something wrong.
	 */
	if (init_global_variables () != 0)
		return (1);

	if (test_config)
		return (0);

#if COLLECT_DAEMON
	/*
	 * fork off child
	 */
	memset (&sig_chld_action, '\0', sizeof (sig_chld_action));
	sig_chld_action.sa_handler = SIG_IGN;
	sigaction (SIGCHLD, &sig_chld_action, NULL);

	if (daemonize)
	{
		if ((pid = fork ()) == -1)
		{
			/* error */
			char errbuf[1024];
			fprintf (stderr, "fork: %s",
					sstrerror (errno, errbuf,
						sizeof (errbuf)));
			return (1);
		}
		else if (pid != 0)
		{
			/* parent */
			/* printf ("Running (PID %i)\n", pid); */
			return (0);
		}

		/* Detach from session */
		setsid ();

		/* Write pidfile */
		if (pidfile_create ())
			exit (2);

		/* close standard descriptors */
		close (2);
		close (1);
		close (0);

		if (open ("/dev/null", O_RDWR) != 0)
		{
			ERROR ("Error: Could not connect `STDIN' to `/dev/null'");
			return (1);
		}
		if (dup (0) != 1)
		{
			ERROR ("Error: Could not connect `STDOUT' to `/dev/null'");
			return (1);
		}
		if (dup (0) != 2)
		{
			ERROR ("Error: Could not connect `STDERR' to `/dev/null'");
			return (1);
		}
	} /* if (daemonize) */
#endif /* COLLECT_DAEMON */

	memset (&sig_pipe_action, '\0', sizeof (sig_pipe_action));
	sig_pipe_action.sa_handler = SIG_IGN;
	sigaction (SIGPIPE, &sig_pipe_action, NULL);

	/*
	 * install signal handlers
	 */
	memset (&sig_int_action, '\0', sizeof (sig_int_action));
	sig_int_action.sa_handler = sig_int_handler;
	if (0 != sigaction (SIGINT, &sig_int_action, NULL)) {
		char errbuf[1024];
		ERROR ("Error: Failed to install a signal handler for signal INT: %s",
				sstrerror (errno, errbuf, sizeof (errbuf)));
		return (1);
	}

	memset (&sig_term_action, '\0', sizeof (sig_term_action));
	sig_term_action.sa_handler = sig_term_handler;
	if (0 != sigaction (SIGTERM, &sig_term_action, NULL)) {
		char errbuf[1024];
		ERROR ("Error: Failed to install a signal handler for signal TERM: %s",
				sstrerror (errno, errbuf, sizeof (errbuf)));
		return (1);
	}

	memset (&sig_usr1_action, '\0', sizeof (sig_usr1_action));
	sig_usr1_action.sa_handler = sig_usr1_handler;
	if (0 != sigaction (SIGUSR1, &sig_usr1_action, NULL)) {
		char errbuf[1024];
		ERROR ("Error: Failed to install a signal handler for signal USR1: %s",
				sstrerror (errno, errbuf, sizeof (errbuf)));
		return (1);
	}

	/*
	 * run the actual loops
	 */
	do_init ();

	if (test_readall)
	{
		if (plugin_read_all_once () != 0)
			exit_status = 1;
	}
	else
	{
		INFO ("Initialization complete, entering read-loop.");
		do_loop ();
	}

	/* close syslog */
	INFO ("Exiting normally.");

	do_shutdown ();

#if COLLECT_DAEMON
	if (daemonize)
		pidfile_remove ();
#endif /* COLLECT_DAEMON */

	return (exit_status);
} /* int main */
